bpf/lib/fib.h

```diff
@@ -0,0 +1,120 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* Copyright (C) 2020 Authors of Cilium */
+
+#ifndef __LIB_FIB_H_
+#define __LIB_FIB_H_
+
+#include <bpf/ctx/ctx.h>
+#include <bpf/api.h>
+
+#include "common.h"
+#include "l3.h"
+
+#ifdef ENABLE_IPV6
+static __always_inline int
+redirect_direct_v6(struct __ctx_buff *ctx __maybe_unused,
+		   int l3_off __maybe_unused,
+		   struct ipv6hdr *ip6 __maybe_unused)
+{
+# ifdef ENABLE_REDIRECT_FAST
+	int ret, nh_size = 0, oif = DIRECT_ROUTING_DEV_IFINDEX;
+	struct bpf_redir_neigh *nh = NULL;
+	bool no_neigh = true;
+#  ifndef ENABLE_SKIP_FIB
+	struct bpf_redir_neigh nh_params;
+	struct bpf_fib_lookup fib_params = {
+		.family		= AF_INET6,
+		.ifindex	= ctx->ingress_ifindex,
+	};
+
+	ipv6_addr_copy((union v6addr *)&fib_params.ipv6_src,
+		       (union v6addr *)&ip6->saddr);
+	ipv6_addr_copy((union v6addr *)&fib_params.ipv6_dst,
+		       (union v6addr *)&ip6->daddr);
+
+	ret = fib_lookup(ctx, &fib_params, sizeof(fib_params),
+			 BPF_FIB_LOOKUP_DIRECT);
+	if (unlikely(ret == BPF_FIB_LKUP_RET_NOT_FWDED || ret < 0))
+		return CTX_ACT_DROP;
+	if (ret == BPF_FIB_LKUP_RET_NO_NEIGH) {
+		nh_params.nh_family = fib_params.family;
+		__bpf_memcpy_builtin(&nh_params.ipv6_nh, &fib_params.ipv6_dst,
+				     sizeof(nh_params.ipv6_nh));
+		no_neigh = true;
+		nh = &nh_params;
+		nh_size = sizeof(*nh);
+	}
+
+	oif = fib_params.ifindex;
+#  endif /* ENABLE_SKIP_FIB */
+
+	ret = ipv6_l3(ctx, l3_off, NULL, NULL, METRIC_EGRESS);
+	if (unlikely(ret != CTX_ACT_OK))
+		return ret;
+	if (no_neigh)
+		return redirect_neigh(oif, nh, nh_size, 0);
+#  ifndef ENABLE_SKIP_FIB
+	if (eth_store_daddr(ctx, fib_params.dmac, 0) < 0)
+		return CTX_ACT_DROP;
+	if (eth_store_saddr(ctx, fib_params.smac, 0) < 0)
+		return CTX_ACT_DROP;
+	return redirect(oif, 0);
+#  endif /* ENABLE_SKIP_FIB */
+# endif /* ENABLE_REDIRECT_FAST */
+	return CTX_ACT_DROP;
+}
+#endif /* ENABLE_IPV6 */
+
+#ifdef ENABLE_IPV4
+static __always_inline int
+redirect_direct_v4(struct __ctx_buff *ctx __maybe_unused,
+		   int l3_off __maybe_unused,
+		   struct iphdr *ip4 __maybe_unused)
+{
+# ifdef ENABLE_REDIRECT_FAST
+	int ret, nh_size = 0, oif = DIRECT_ROUTING_DEV_IFINDEX;
+	struct bpf_redir_neigh *nh = NULL;
+	bool no_neigh = true;
```

Comment:
Shouldn't this be set to false? Otherwise, I don't see how this can become non-true.

Reply:
indeed, good catch, it should be `no_neigh = !is_defined(EMABLE_SKIP_FIB)`